include/util.h

```diff
@@ -1,10 +1,34 @@
 #pragma once
 
+typedef long long int ll;
+typedef unsigned long long int ull;
+
 /**
  * Cast a double to unsigned long long with epsilon adjustment.
  * @param d         the double to cast.
  * @param epsilon   optional parameter representing the epsilon to use.
  */
-unsigned long long int double_to_ull(double d, double epsilon = 0.00000001) {
-  return (unsigned long long) (d + epsilon);
-}
+unsigned long long int double_to_ull(double d, double epsilon = 0.00000001);
```

Comment:
I still don't quite understand what this is meant to do (or rather, why std::round is insufficient)

Reply:
Admittedly, this is a hack to get around another, hackier, hack. We want the same effect as integer flooring, but there's a weird quirk in floating point to unsigned types where sometimes a value very close to a whole number `n` will round down to `n-1`. We could compromise by always rounding up, but since this rounding shows up in creating `c*log n` buckets, this may end up increasing our memory footprint by a non-insignificant amount.

To me, this is the least-bad option, though I would like to hear your thoughts on it.

Reply:
I guess I'm asking why std::round wouldn't work specifically - it should always round to the nearest integer (sometimes up, sometimes down) except in cases where that integer is unrepresentable, in which case it will round to the nearest representable int (which I don't think we should be hitting)

Reply:
std::round would work to give us something that could work, but we may run into an issue where it causes us to use more memory than we need to and perhaps run longer than we need to. In almost every other situation I would agree with using std::round, but since this is an area where we need to be very parsimonious with our memory usage, I'm inclined against using a round.